data_structures/linked_list.cpp

```diff
@@ -1,134 +1,240 @@
+ */
+bool isDigit(const std::string &s) {
+    // function statements here
+    for (char i : s) {
+        if (!isdigit(i)) {
+            return false;
         }
-        t->next = n;
-    } else {
-        start = n;
     }
+    return true;
 }
 
-void remove(int x) {
-    if (start == NULL) {
-        std::cout << "\nLinked List is empty\n";
-        return;
-    } else if (start->val == x) {
-        node *temp = start;
-        start = start->next;
-        delete temp;
-        return;
-    }
+/**
+ * A link class containing a value and pointer to another link
+ */
+class link {
+ public:
+    explicit link(int v = 0, link *s = nullptr);
 
-    node *temp = start, *parent = start;
+    std::shared_ptr<link> succ;  ///< pointer to the next value on the list
+    int val;                     ///< value of the current link
+};
 
-    while (temp != NULL && temp->val != x) {
-        parent = temp;
-        temp = temp->next;
-    }
+/**
+ * constructor of the class link
+ * @param v is used to initialize class member value
+ * @param s is used to initilaize class member succ
+ */
+link::link(int v, link *s) : val(v), succ(s) {}
 
-    if (temp == NULL) {
-        std::cout << std::endl << x << " not found in list\n";
-        return;
+/**
+ * A list class containing a sequence of links
+ */
+class list {
+ public:
+    list() : first(new link), last(first) {}
+
+    bool isEmpty();
+
+    void push_back(int new_elem);
+    void push_front(int new_elem);
+    void erase(int old_elem);
+    void display();
+    void search(int find_elem);
+    void reverse();
+
+ private:
+    std::shared_ptr<link> first;  ///< link before the actual first element
+    std::shared_ptr<link> last;   ///< last link on the list
+};
+
+/**
+ * function checks if list is empty
+ * @returns true if list is empty
+ * @returns false if list is not empty
+ */
+bool list::isEmpty() {
+    if (first == last) {
+        return true;
+    } else {
+        return false;
     }
+}
 
-    parent->next = temp->next;
-    delete temp;
+/**
+ * function adds new element to the end of the list
+ * @param new_elem to be added to the end of the list
+ */
+void list::push_back(int new_elem) {
+    std::shared_ptr<link> new_link(new link(new_elem));
+    last->succ = new_link;
+    last = last->succ;
 }
 
-void search(int x) {
-    node *t = start;
-    int found = 0;
-    while (t != NULL) {
-        if (t->val == x) {
-            std::cout << "\nFound";
-            found = 1;
-            break;
-        }
-        t = t->next;
+/**
+ * function adds new element to the beginning of the list
+ * @param new_elem to be added to front of the list
+ */
+void list::push_front(int new_elem) {
+    std::shared_ptr<link> new_link(new link(new_elem));
+    new_link->succ = first->succ;
+    first->succ = new_link;
+}
+
+/**
+ * function erases old element from the list
+ * @param old_elem to be erased from the list
+ */
+void list::erase(int old_elem) {
+    if (isEmpty()) {
```

Comment:
to erase a shared pointer, use [`std::shared_ptr<T>::reset()`](https://en.cppreference.com/w/cpp/memory/shared_ptr/reset)